documentation.md

```diff
@@ -0,0 +1,96 @@
+    participant RedisCache as Redis Cache
+    participant Database as Database
+    participant CMS as CMS
+    participant CADP as CADP
+
+    Note over Caller, HandleTransaction: The process starts by handling a transaction.
+
+    Caller->>HandleTransaction: call(transaction, networkMap, ruleResult)
+    HandleTransaction->>+ExecuteRequest: call(transaction, typology, ruleResult, networkMap)
+
+    Note over ExecuteRequest: Execute Request processes the transaction for each typology.
+
+    loop channel.typologies
+        ExecuteRequest->>RedisCache: addOneGetAll(cacheKey, ruleResult)
+        ExecuteRequest->>RedisCache: check cached ruleResults
+
+        Note over RedisCache: Redis Cache stores interim rule results for faster processing.
+
+        opt no cached ruleResults
+            ExecuteRequest->>HandleTransaction: return CADPRequest
+        end
+        ExecuteRequest->>Database: getTypologyExpression(typology)
+
+        Note over Database: Database provides the typology expression needed for evaluation.
+
+        ExecuteRequest->>EvaluateTypologyExpr: call(ruleValues, ruleResults, typologyExpression)
+
+        Note over EvaluateTypologyExpr: Evaluate Typology Expr computes the typology result value.
+
+        opt typologyResultValue > expression.threshold
+            ExecuteRequest->>CMS: send Typology result
+        end
+
+        Note over CMS: CMS receives typology results that exceed the threshold.
+
+        ExecuteRequest->>CADP: send Typology result
+
+        Note over CADP: CADP receives all the typology results.
+
+        ExecuteRequest->>RedisCache: deleteKey(cacheKey)
+
+        Note over RedisCache: Redis Cache deletes the interim cache key after processing.
+
+        ExecuteRequest->>HandleTransaction: return CADPRequest
+    end
+
+    HandleTransaction-->>ExecuteRequest: CADPRequest returned
+    HandleTransaction->>HandleTransaction: add CADPRequest to CombinedResult.cadpRequests
+
+    Note over HandleTransaction: Handle Transaction accumulates the CADPRequests.
+
+    HandleTransaction->>Caller: return CombinedResult
+
+    Note over Caller: The final combined result is returned to the caller.
+
+
+```
+
+Here's a detailed explanation of the numbers in the sequence diagram:
+
+1. The process starts by handling a transaction. The caller initiates the request to process a transaction by calling the HandleTransaction function with the transaction, network map, and rule result as input parameters.
+2. The HandleTransaction function, in turn, calls the ExecuteRequest function for each typology in the transaction. It passes the transaction, typology, rule result, and network map as input parameters.
+3. The ExecuteRequest function processes the transaction for each typology. It interacts with Redis Cache to store and retrieve interim rule results for faster processing.
+4. If there are no cached rule results, the ExecuteRequest function returns a CADPRequest to the HandleTransaction function.
```

Comment:
If I am reading step 3 correctly in the diagram (addOneGetAll), the impression I get here is that the interaction with the cache deviates slightly from the original specification. In the user story, the cache is read and checked, and if the typology is not complete, the rule is then added to the cache. In this diagram, based on the name of the function, it looks like the incoming rule is immediately cached and then the cache is read. @foleyfrms - can you confirm?

While this is not necessarily wrong (from the perspective of efficiency, we're only hitting the cache once, and we are going to be updating the cache more frequently than not, so the "savings" from not writing that cache when the final rule result comes in is probably negligible), I do not understand how the "exit condition" [no cached results] as it is worded is possible. If we are updating the cache with the rule result before step 5, then there will ALWAYS be at least one rule result in the cache at the time that the evaluation takes place during step 4.

Do we need to reword the evaluation to [typology incomplete]? Or is this some kind of "error" condition that indicates that something went wrong?

Reply:
Yes we modified the way we interact with cache in this instance - to add and get the result back in a single call to redis, as the method name indicates (part of the process to lower the amount of chatter generated in E2E). 

The outcome here mentioned ("no cached results") doesn't exist in the code - instead, the only exit condition here is when the amount of results returned from cache < amount of rules in network map for this typology, eg, indicating that the typology has not yet received all the rule results it's expecting.